lib/iris/util.py

```diff
@@ -1593,6 +1593,173 @@ def demote_dim_coord_to_aux_coord(cube, name_or_coord):
+                  y                           x
+                  latitude                    x
+                  longitude                   x
+    """
+    import numpy as np
+    from iris.cube import Cube
+    from iris.coords import AuxCoord
+
+    all_dims = range(cube.ndim)
+
+    dims = dims if dims is not None else all_dims
+
+    other_dims = list(set(all_dims) - set(dims))
+    shape_array = np.asarray(cube.shape)
+    # The new (flat) dimension will be in the first dimension of the aux coord
+    flat_dim = dims[0]
+    dim_shape = shape_array[list(dims)]
+    new_dim_shape = np.product(dim_shape)
+    new_shape = np.insert(shape_array[other_dims], flat_dim, new_dim_shape)
+
+    # Figure out the new dimensions (so we know where to put the dim-coords)
+    new_dims = np.zeros_like(shape_array, dtype=int)
+    new_dims[other_dims] = np.arange(0, len(other_dims))
+    new_dims[flat_dim:] += 1
+
+    # Get all the coords which span the dimensions to be flattened
+    coords_to_flatten = cube.coords(dimensions=dims)
+    # And any other coords which span any one of the dimensions
+    # These (1-D) coordinates are expanded before being flattened
+    coords_to_ignore = []
+    for aux_dim, coord_dim in enumerate(dims):
+        # Expand the coords which need expanding, then add to flatten list
+        for c in cube.coords(dimensions=coord_dim):
+            # I have to add these here separately since they change
+            coords_to_ignore.append(c)
+            # Broadcast the coord (which must be one-dimensional)
+            new_points = broadcast_to_shape(c.points, dim_shape, (aux_dim,))
+            new_bounds = None if c.bounds is None else \
+                broadcast_to_shape(c.bounds, list(dim_shape) + [2],
+                                   (aux_dim, -1))
+            coords_to_flatten.append(
+                AuxCoord.from_coord(c).copy(new_points, new_bounds))
+
+    other_dim_coords = [(c, new_dims[cube.coord_dims(c)]) for c in
+                        cube.coords(dim_coords=True)
+                        if c not in coords_to_flatten and
+                        c not in coords_to_ignore]
+    other_aux_coords = [(c, cube.coord_dims(c)) for c in
+                        cube.coords(dim_coords=False)
+                        if c not in coords_to_flatten and
+                        c not in coords_to_ignore and
+                        c not in cube.derived_coords]
+
+    new_data = cube.core_data().reshape(new_shape)
+    new_aux_coords = other_aux_coords
+    for c in coords_to_flatten:
+        # The new (flat) AuxCoords are always the last dimension
+        new_bounds = None if c.bounds is None else c.bounds.reshape(-1, 2)
+        # Any DimCoords spanning these dimensions have been demoted so
+        # there should only be AuxCoords left
+        new_aux_coords.append((c.copy(c.points.flat, new_bounds), flat_dim))
+    # TODO: Deal with Aux Factories and cell measures....
```

Comment:
As stated, I think these aspects need dealing with, or at least to detect and issue an error.